tools/test/test_depsolve.py

```diff
@@ -0,0 +1,254 @@
+import configparser
+import importlib
+import json
+import os
+import socket
+import subprocess as sp
+from tempfile import TemporaryDirectory
+
+import pytest
+
+REPO_PATHS = [
+    "./test/data/testrepos/baseos/",
+    "./test/data/testrepos/custom/",
+]
+
+# osbuild-depsolve-dnf uses the GPG header to detect if keys are defined in-line or as file paths/URLs
+TEST_KEY = "-----BEGIN PGP PUBLIC KEY BLOCK-----\nTEST KEY\n"
+
+
+def has_dnf5():
+    return bool(importlib.util.find_spec("libdnf5"))
+
+
+def depsolve(pkgs, repos, root_dir, cache_dir, command):
+    req = {
+        "command": "depsolve",
+        "arch": "x86_64",
+        "module_platform_id": "platform:el9",
+        "releasever": "9",
+        "cachedir": cache_dir,
+        "arguments": {
+            "root_dir": root_dir,
+            "repos": repos,
+            "transactions": [
+                {"package-specs": pkgs},
+            ]
+        }
+    }
+    p = sp.run([command], input=json.dumps(req).encode(), check=True, capture_output=True)
+    if p.stderr:
+        print(p.stderr.decode())
+
+    return json.loads(p.stdout.decode())
+
+
+def get_rand_port():
+    s = socket.socket()
+    s.bind(("", 0))
+    return s.getsockname()[1]
+
+
+@pytest.fixture(name="repo_servers", scope="module")
+def repo_servers_fixture():
+    procs = []
+    addresses = []
+    for path in REPO_PATHS:
+        port = get_rand_port()  # this is racy, but should be okay
+        p = sp.Popen(["python3", "-m", "http.server", str(port)], cwd=path, stdout=sp.PIPE)
```

Comment:
There is a small race here, if the server takes some time to be ready but we (async) moved on and query the repo before the server is fully up ths will error. Not a big deal, we should probably just fix in a followup (and move them all into a httpd_server_dir() helper or something).